velox/exec/Merge.cpp

```diff
@@ -521,6 +435,116 @@ bool SourceStream::fetchMoreData(std::vector<ContinueFuture>& futures) {
+      batchStreams_(createBatchStreams(std::move(spillReadFilesGroup))),
+      sourceMerger_(createSourceMerger(
+          sortingKeys,
+          type,
+          outputBatchSize,
+          sources_,
+          pool)),
+      executor_(executor),
+      pool_(pool) {
+  if (executor_ == nullptr) {
+    kExecutor_ = std::make_shared<folly::IOThreadPoolExecutor>(
+        std::thread::hardware_concurrency());
+    executor_ = kExecutor_.get();
+  }
+  createFileStreamProducers(batchStreams_, sources_);
+}
+
+RowVectorPtr SpillMerger::getOutput(
+    std::vector<ContinueFuture>& sourceBlockingFutures,
+    bool& atEnd) const {
+  sourceMerger_->isBlocked(sourceBlockingFutures);
+  if (!sourceBlockingFutures.empty()) {
+    return nullptr;
+  }
+  return sourceMerger_->getOutput(sourceBlockingFutures, atEnd);
+}
+
+std::vector<std::shared_ptr<MergeSource>> SpillMerger::createMergeSources(
+    int num) {
+  std::vector<std::shared_ptr<MergeSource>> sources;
+  sources.reserve(num);
+  for (auto i = 0; i < num; ++i) {
+    sources.push_back(MergeSource::createLocalMergeSource());
+  }
+  for (const auto& source : sources) {
+    source->start();
+  }
+  return sources;
+}
+
+std::vector<std::unique_ptr<BatchStream>> SpillMerger::createBatchStreams(
+    std::vector<std::vector<std::unique_ptr<SpillReadFile>>>
+        spillReadFilesGroup) {
+  const auto numStreams = spillReadFilesGroup.size();
+  std::vector<std::unique_ptr<BatchStream>> batchStreams;
+  batchStreams.reserve(numStreams);
+  for (auto i = 0; i < numStreams; ++i) {
+    batchStreams.emplace_back(
+        ConcatFilesSpillBatchStream::create(std::move(spillReadFilesGroup[i])));
+  }
+  return batchStreams;
+}
+
+std::unique_ptr<SourceMerger> SpillMerger::createSourceMerger(
+    const std::vector<SpillSortKey>& sortingKeys,
+    const RowTypePtr& type,
+    vector_size_t outputBatchSize,
+    const std::vector<std::shared_ptr<MergeSource>>& sources,
+    velox::memory::MemoryPool* pool) {
+  std::vector<std::unique_ptr<SourceStream>> cursors;
+  cursors.reserve(sources.size());
+  for (auto i = 0; i < sources.size(); ++i) {
+    cursors.push_back(std::make_unique<SourceStream>(
+        sources[i].get(), sortingKeys, outputBatchSize));
+  }
+  return std::make_unique<SourceMerger>(
+      type, outputBatchSize, std::move(cursors), pool);
+}
+
+folly::Future<folly::Unit> SpillMerger::fileStreamProducer(
+    BatchStream* batchStream,
+    MergeSource* source) const {
+  RowVectorPtr vector;
+  ContinueFuture future;
+  if (!batchStream->nextBatch(vector)) {
+    source->enqueue(nullptr, &future);
+    return folly::makeFuture();
+  }
+
+  source->enqueue(std::move(vector), &future);
+  return std::move(future)
+      .via(executor_)
+      .thenValue([this, batchStream, source](folly::Unit) {
+        fileStreamProducer(batchStream, source);
+      })
+      .thenError(folly::tag_t<std::exception>{}, [](const std::exception& e) {
+        VELOX_FAIL(e.what());
```

Comment:
Just log error message from the exception from the driver and stop the continuation of batch streaming